src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs

```diff
@@ -56,149 +67,301 @@ public sealed class RichTextEditorPastedImages
+        }
+
+        foreach (HtmlNode? img in tmpImages)
+        {
+            // The data attribute contains the path to the tmp img to persist as a media item
+            var tmpImgPath = img.GetAttributeValue(TemporaryImageDataAttribute, string.Empty);
+
+            if (string.IsNullOrEmpty(tmpImgPath))
             {
-                throw new PanicException(
-                    $"Could not find media by id {udi.Guid} or there was no UmbracoContext available.");
+                continue;
             }
 
-            var location = mediaTyped.Url(_publishedUrlProvider);
+            var qualifiedTmpImgPath = _hostingEnvironment.MapPathContentRoot(tmpImgPath);
+
+            PersistMediaItem(mediaParentFolder, userId, img, qualifiedTmpImgPath);
+        }
+
+        return htmlDoc.DocumentNode.OuterHtml;
+    }
+
+    private void PersistMediaItem(Guid mediaParentFolder, int userId, HtmlNode img, string qualifiedTmpImgPath)
+    {
+        var absoluteTempImagePath = Path.GetFullPath(qualifiedTmpImgPath);
+
+        if (IsValidPath(absoluteTempImagePath) == false)
+        {
+            return;
+        }
+
+        var fileName = Path.GetFileName(absoluteTempImagePath);
+        var safeFileName = fileName.ToSafeFileName(_shortStringHelper);
 
-            // Find the width & height attributes as we need to set the imageprocessor QueryString
-            var width = img.GetAttributeValue("width", int.MinValue);
-            var height = img.GetAttributeValue("height", int.MinValue);
+        var mediaItemName = safeFileName.ToFriendlyName();
+        GuidUdi udi;
+
+        if (_uploadedImages.ContainsKey(qualifiedTmpImgPath) == false)
+        {
+            var isSvg = qualifiedTmpImgPath.EndsWith(".svg");
+            var mediaType = isSvg
+                ? Constants.Conventions.MediaTypes.VectorGraphicsAlias
+                : Constants.Conventions.MediaTypes.Image;
 
-            if (width != int.MinValue && height != int.MinValue)
+            IMedia mediaFile;
+            if (mediaParentFolder == Guid.Empty)
             {
-                location = imageUrlGenerator.GetImageUrl(new ImageUrlGenerationOptions(location)
-                {
-                    ImageCropMode = ImageCropMode.Max,
-                    Width = width,
-                    Height = height,
-                });
+                mediaFile = _mediaService.CreateMedia(mediaItemName, Constants.System.Root, mediaType, userId);
+            }
+            else
+            {
+                mediaFile = _mediaService.CreateMedia(mediaItemName, mediaParentFolder, mediaType, userId);
             }
 
-            img.SetAttributeValue("src", location);
+            var fileInfo = new FileInfo(absoluteTempImagePath);
 
-            // Remove the data attribute (so we do not re-process this)
-            img.Attributes.Remove(TemporaryImageDataAttribute);
+            FileStream? fileStream = fileInfo.OpenReadWithRetry();
+            if (fileStream == null)
```

Comment:
Lets change this to pattern matching while we're at it 🔥
```suggestion
            if (fileStream is null)
```